File: elixir/bob/lib/bob.ex
defmodule Bob do
  def hey(input) do
    case String.trim(input) do
      "" ->
        "Fine. Be that way!"

      remark ->
        respond_to_verbal_remark(remark)
    end
  end

  defp respond_to_verbal_remark(remark) do
    case {question?(remark), yelling?(remark)} do
      {true, true} ->
        "Calm down, I know what I'm doing!"

      {true, false} ->
        "Sure."

      {false, true} ->
        "Whoa, chill out!"

      {false, false} ->
        "Whatever."
    end
  end

  defp question?(remark), do: String.ends_with?(remark, "?")

  defp yelling?(remark) do
    remark == String.upcase(remark) and remark != String.downcase(remark)
  end
end

File: java/bob/src/main/java/Bob.java
class Bob {
    private static final String LETTER_INPUT = ".*[A-Za-z].*";

    String hey(String input) {
        String remark = input.trim();

        if (isSilence(remark)) {
            return "Fine. Be that way!";
        } else {
            return respondToVerbalRemark(remark);
        }
    }

    private boolean isSilence(String remark) {
        return remark.isEmpty();
    }

    private String respondToVerbalRemark(String remark) {
        boolean question = isQuestion(remark);
        boolean yelling = isYelling(remark);

        if (question && yelling) {
            return "Calm down, I know what I'm doing!";
        } else if (question) {
            return "Sure.";
        } else if (yelling) {
            return "Whoa, chill out!";
        } else {
            return "Whatever.";
        }
    }

    private boolean isQuestion(String remark) {
        return remark.endsWith("?");
    }

    private boolean isYelling(String remark) {
        String yelling = remark.toUpperCase();
        return hasLetters(remark) && yelling.equals(remark);
    }

    private boolean hasLetters(String remark) {
        return remark.matches(LETTER_INPUT);
    }
}
